File: packages/api/pixl_api/schemas/workflows.py
"""Pydantic request/response models for workflow endpoints."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel


class WorkflowSummary(BaseModel):
    id: str
    name: str
    description: str | None = None
    version: str | None = None
    tags: list[str] | None = None
    tier: str | None = None
    routing: dict[str, Any] | None = None


class WorkflowNodeDetail(BaseModel):
    id: str
    type: str
    task_config: dict[str, Any] | None = None
    gate_config: dict[str, Any] | None = None
    hook_config: dict[str, Any] | None = None
    metadata: dict[str, str] | None = None


class WorkflowEdgeDetail(BaseModel):
    to: str
    on: str
    condition: str | None = None


class WorkflowLoopDetail(BaseModel):
    id: str
    from_node: str
    to_node: str
    max_iterations: int | str = 3
    edge_trigger: str = "failure"


class WorkflowDetail(BaseModel):
    id: str
    name: str
    description: str | None = None
    tags: list[str] | None = None
    version: str | None = None
    tier: str | None = None
    routing: dict[str, Any] | None = None
    nodes: dict[str, WorkflowNodeDetail]
    edges: dict[str, list[WorkflowEdgeDetail]]
    loops: list[WorkflowLoopDetail] | None = None
    stages: list[dict[str, Any]] | None = None
    variables: dict[str, str] | None = None

File: packages/console/src/types/workflows.ts
export interface WorkflowRoutingHints {
  auto_route: boolean;
  sub_invocable: boolean;
  category: string;
  trigger_keywords: string[];
}

export interface WorkflowSummary {
  id: string;
  name: string;
  description?: string;
  version?: string;
  tags?: string[];
  source?: string;
  tier?: string;
  routing?: WorkflowRoutingHints;
  [key: string]: unknown;
}

export interface WorkflowNodeDetail {
  id: string;
  type: "task" | "gate" | "hook" | "parallel" | "merge" | "sub_workflow";
  task_config: {
    agent: string;
    max_turns: number;
    retry_policy?: unknown;
  } | null;
  gate_config: {
    id: string;
    name: string;
    description: string;
    timeout_minutes: number | null;
    timeout_policy: string;
  } | null;
  hook_config: { hook: string } | null;
  metadata?: Record<string, string>;
}

export interface WorkflowEdgeDetail {
  to: string;
  on: string;
  condition: string | null;
}

export interface WorkflowLoopDetail {
  id: string;
  from_node: string;
  to_node: string;
  max_iterations: number;
  edge_trigger: string;
}

export interface WorkflowDetail {
  id: string;
  name: string;
  description: string;
  tags: string[];
  version: string;
  tier?: string;
  routing?: WorkflowRoutingHints;
  nodes: Record<string, WorkflowNodeDetail>;
  edges: Record<string, WorkflowEdgeDetail[]>;
  loops: WorkflowLoopDetail[];
  stages: unknown[];
  variables: Record<string, string>;
}
